### PR TITLE
count request and add limit to RECONNECT, process room-leave event for bot been removed.

### PR DESCRIPTION
**Pull request by su-chang**
## I'm submitting a...

```
[x] Bug Fix
[ ] Feature
[ ] Other (Refactoring, Added tests, Documentation, ...)
```

## Checklist

- [ ] Commit Messages follow the [Conventional Commits](https://conventionalcommits.org/) pattern
  - A feature commit message is prefixed "feat:"
  - A bugfix commit message is prefixed "fix:"
- [ ] Tests for the changes have been added

## Description

_please describe the changes that you are making_

_for features, please describe how to use the new feature_

_please include a reference to an existing issue, if applicable_

## Does this PR introduce a breaking change?

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/server-manager/grpc-gateway.ts [hunk 1]
```diff
@@ -217,29 +221,54 @@ export class GrpcGateway extends EventEmitter {
     request.setRequestid(requestId)
 
     try {
+      if (this.timeoutNumber > 10 && Date.now() - this.startTime <= 3 * 60 * 1000) {
```
Comment (windmemory) [inline src/server-manager/grpc-gateway.ts]: We shouldn't put the check at the beginning of the request, we should do a reconnect when we found the last timeout happened.

File: src/padplus-manager/padplus-manager.ts [hunk 2]
```diff
@@ -591,8 +591,10 @@ export class PadplusManager extends EventEmitter {
             const deleteUserName = contactData.field
             if (this.cacheManager) {
               if (isRoomId(deleteUserName)) {
-                await this.cacheManager.deleteRoomMember(deleteUserName)
-                await this.cacheManager.deleteRoom(deleteUserName)
+                setTimeout(async () => {
+                  await this.cacheManager!.deleteRoomMember(deleteUserName)
```
Comment (windmemory) [inline src/padplus-manager/padplus-manager.ts]: Would a flag be better here to mark the room as deleted?

Comment (su-chang) [reply src/padplus-manager/padplus-manager.ts]: Done!

File: src/server-manager/grpc-gateway.ts (hunk 1 above)
Comment (su-chang) [reply src/server-manager/grpc-gateway.ts]: Done!

File: src/server-manager/cache-manager.ts [hunk 3]
```diff
@@ -59,7 +59,7 @@ export class CacheManager {
   private cacheWXID?    : FlashStore<string, string>
   private cacheContactRawPayload?    : FlashStore<string, PadplusContactPayload>
   private cacheRoomMemberRawPayload? : FlashStore<string, {
-    [contactId: string]: PadplusRoomMemberPayload,
+    [contactId: string]: PadplusRoomMemberPayload | boolean,
```
Comment (windmemory) [inline src/server-manager/cache-manager.ts]: why add the `boolean` here?

File: src/server-manager/cache-manager.ts [hunk 4]
```diff
@@ -223,7 +223,7 @@ export class CacheManager {
    */
   public async getRoomMember (
     roomId: string,
-  ): Promise<{ [contactId: string]: PadplusRoomMemberPayload } | undefined> {
+  ): Promise<{ [contactId: string]: PadplusRoomMemberPayload | boolean } | undefined> {
```
Comment (windmemory) [inline src/server-manager/cache-manager.ts]: ditto

File: src/server-manager/cache-manager.ts [hunk 5]
```diff
@@ -232,7 +232,7 @@ export class CacheManager {
 
   public async setRoomMember (
     roomId: string,
-    payload: { [contactId: string]: PadplusRoomMemberPayload }
+    payload: { [contactId: string]: PadplusRoomMemberPayload | boolean }
```
Comment (windmemory) [inline src/server-manager/cache-manager.ts]: ditto

File: src/server-manager/grpc-gateway.ts [hunk 6]
```diff
@@ -204,6 +208,23 @@ export class GrpcGateway extends EventEmitter {
...
+  private async checkTimeout (uin: string) {
+    this.timeoutNumber++
+    if (this.timeoutNumber > 10 && Date.now() - this.startTime <= 3 * 60 * 1000) {
+      await this.request(
+        ApiType.RECONNECT,
+        uin,
+      )
+      this.startTime = Date.now()
+      this.timeoutNumber = 0
+    } else if (this.timeoutNumber === 0) {
```
Comment (windmemory) [inline src/server-manager/grpc-gateway.ts]: this line will never be executed.

File: src/padplus-manager/padplus-manager.ts [hunk 7]
```diff
@@ -591,8 +591,15 @@ export class PadplusManager extends EventEmitter {
             const deleteUserName = contactData.field
             if (this.cacheManager) {
               if (isRoomId(deleteUserName)) {
-                await this.cacheManager.deleteRoomMember(deleteUserName)
-                await this.cacheManager.deleteRoom(deleteUserName)
+                const room = await this.cacheManager.getRoom(deleteUserName)
+                const roomMember = await this.cacheManager.getRoomMember(deleteUserName)
+                if (!room || !roomMember) {
```
Comment (windmemory) [inline src/padplus-manager/padplus-manager.ts]: maybe we should insert a record with only `id` and `delated` flag into the cache here?

**Pull request merged**